File: UserInterface/MainForm.cs
/*
 * Kevin Nguyen
 */ 
using System;
using System.Windows.Forms;
using CommonFramework.Core;
using TheBrain;
// Kevin Nguyen
namespace UserInterface
{
    public partial class MainForm : Form
    {
        public MainForm()
        {
            InitializeComponent();
        }
        ///<summary>
        ///a local field (variable to hold the connection / link to the DLL's class
        ///This line is said to instantiate the class.
        /// </summary>
        private TicTacToeGame _ticTacToeGame = new TicTacToeGame(); 

        private void label1_Click(object sender, EventArgs e)
        {

        }

        private void MainForm_Load(object sender, EventArgs e)
        {
            EnableControls(false);
            _ticTacToeGame.MarkerChanged += this.MarkerChangedHandler;
        }
        private void btnAbout_Click(object sender, EventArgs e)
        {
            var aboutBox = new AboutBoxTicTacToe();
            aboutBox.ShowDialog();
        }

        private void btnClose_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void btnStartGame_Click(object sender, EventArgs e)
        {
            _ticTacToeGame.ResetGrid();
            foreach (var button in pnlButtons.Controls)
            {
                if (button is Button btn)
                {
                    btn.Text = "?";
                }
            }
            /*we use using the Brain to use references to the DLL located in that cs file
             * we create a local varaible with class TicTacToeGame from theBrain cs file and name it _tictacToeGame
             * since _ticTacToeGame is of class TicTacToe, I can use the method Reset Grid found in the brain's TicTacToe class
             * 
             */ 
        }

        private void btnGoComputer_Click(object sender, EventArgs e)
        {
            if (_ticTacToeGame.Winner != CellOwners.Open) return;
            if (_ticTacToeGame.Winner == CellOwners.Open)
            {
                _ticTacToeGame.AutoPlayComputer();
                EvaluateAndAnnounceWinner();
            }
        }

        void EvaluateAndAnnounceWinner()
        {
            if (_ticTacToeGame.EvaluateBoard())
            {
                MessageBox.Show(
                    $"The Winner is {_ticTacToeGame.IdentifyWinner()}",
                    "Tic Tac Toe Winner",
                    MessageBoxButtons.OK);

            }
        }
        private void txtPlayerName_Validated(object sender, EventArgs e)
        {
            _ticTacToeGame.PlayerName = txtPlayerName.Text;
        }

        private void btnSquare_Common_Click(object sender, EventArgs e)
        {
            if (_ticTacToeGame.Winner != CellOwners.Open) return;
            if (sender is Button btn)
            {
                if (btn.Name.Length == 11 && btn.Name.StartsWith("btnSquare"))
                {
                    var row = btn.Name.Substring(9, 1).ToInt32(-1);
                    var col = btn.Name.Substring(10, 1).ToInt32(-1);
                    _ticTacToeGame.PlaceMarker(row, col, CellOwners.Human);

                    //btn.Text = "X";

                    EvaluateAndAnnounceWinner();
                }
            }
        }
        void EnableControls(bool setControlsEnabled)
        {
            foreach (var control in this.Controls)
            {
                switch (control)
                {
                    case Button btn when (btn.Name != "btnClose" && btn.Name != "btnAbout"):
                        btn.Enabled = setControlsEnabled;
                        break;
                    case Panel pnl:
                        pnl.Enabled = setControlsEnabled;
                        break;
                    default:
                        break;
                }
            }
        }

        private void txtPlayerName_TextChanged(object sender, EventArgs e)
        {
            var nameIsValid = (txtPlayerName.TextLength >= 2);
            EnableControls(nameIsValid);
        }
        private void MarkerChangedHandler(object sender, TheBrain.TicTacToeGame.MarkerChangedArgs e)
        {
            var buttonName = $"btnSquare{e.RowID}{e.ColID}";
            foreach (var control in this.pnlButtons.Controls)
            {
                if (control is Button button)
                {
                    if (button.Name == buttonName)
                    {
                        switch (e.CellOwner)
                        {
                            case CellOwners.Error:
                                button.Text = "#";
                                break;
                            case CellOwners.Open:
                                button.Text = "?";
                                break;
                            case CellOwners.Human:
                                button.Text = "X";
                                break;
                            case CellOwners.Computer:
                                button.Text = "O";
                                break;
                            default:
                                throw new ArgumentOutOfRangeException();
                        }
                    }
                }
            }
        }
    }
}

File: VB Game Surface w Buttons/MainForm.vb

'Kevin Nguyen
Imports TheBrain


Public Class MainForm

    Private _ticTacToeGame As TicTacToeGame = New TicTacToeGame()

    Private Sub Label1_Click(sender As Object, e As EventArgs) Handles lblPlayerName.Click

    End Sub

    Private Sub btnClose_Click(sender As Object, e As EventArgs) Handles btnClose.Click
        Me.Close()
    End Sub

    Private Sub MainForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load

    End Sub

    Private Sub btnAbout_Click(sender As Object, e As EventArgs) Handles btnAbout.Click
        Dim aboutBox As New Form
        aboutBox = New AboutBoxTicTacToe()
        aboutBox.ShowDialog()
    End Sub

    Private Sub btnStartGame_Click(sender As Object, e As EventArgs) Handles btnStartGame.Click
        _ticTacToeGAme.ResetGrid()

        MessageBox.Show("The Start Game button was clicked",
                        "Tic Tac Toe",
                        MessageBoxButtons.OK,
                        MessageBoxIcon.Information)
    End Sub

    Private Sub btnStartComputer_Click(sender As Object, e As EventArgs) Handles btnStartComputer.Click
        MessageBox.Show("The GO Computer button was clicked",
                "Tic Tac Toe",
                MessageBoxButtons.OK,
                MessageBoxIcon.Information)
    End Sub

    Private Sub btnSquare00_Click(sender As Object, e As EventArgs) Handles btnSquare00.Click
        _ticTacToeGame.PlaceMarker(0, 0, CellOwners.Human)
    End Sub

    Private Sub btnSquare01_Click(sender As Object, e As EventArgs) Handles btnSquare01.Click
        _ticTacToeGame.PlaceMarker(0, 1, CellOwners.Human)
    End Sub

    Private Sub btnSquare02_Click(sender As Object, e As EventArgs) Handles btnSquare02.Click
        _ticTacToeGame.PlaceMarker(0, 2, CellOwners.Human)
    End Sub

    Private Sub btnSquare10_Click(sender As Object, e As EventArgs) Handles btnSquare10.Click
        _ticTacToeGame.PlaceMarker(1, 0, CellOwners.Human)
    End Sub

    Private Sub btnSquare11_Click(sender As Object, e As EventArgs) Handles btnSquare11.Click
        _ticTacToeGame.PlaceMarker(1, 1, CellOwners.Human)
    End Sub

    Private Sub btnSquare12_Click(sender As Object, e As EventArgs) Handles btnSquare12.Click
        _ticTacToeGame.PlaceMarker(1, 2, CellOwners.Human)
    End Sub

    Private Sub btnSquare20_Click(sender As Object, e As EventArgs) Handles btnSquare20.Click
        _ticTacToeGame.PlaceMarker(2, 0, CellOwners.Human)
    End Sub

    Private Sub btnSquare21_Click(sender As Object, e As EventArgs) Handles btnSquare21.Click
        _ticTacToeGame.PlaceMarker(2, 1, CellOwners.Human)
    End Sub

    Private Sub btnSquare22_Click(sender As Object, e As EventArgs) Handles btnSquare22.Click
        _ticTacToeGame.PlaceMarker(2, 2, CellOwners.Human)
    End Sub
End Class
